File: code/data_structures/src/stack/sort_stack/sort_stack.js
var Stack = function (initStack = [])
{
    this.stack = initStack;
    this.top = initStack.length;
    
    this.push = function (value) {
        this.stack[this.top++] = value;
    }

    this.size = function() {
        return this.top;
    }

    this.peek = function() {
        return this.stack[this.top - 1];
    }

    this.pop = function () { 
        if (this.top === 0) {
            return "Stack is Empty";
        }
        this.top--;
        let result = this.stack[this.top];
        delete this.stack[this.top];
        return result;
    }

    this.sort = function() {  
        this.stack.sort();
    }

    this.print = function () {  
        console.log(this.stack.join(" "))
    }
}

var myStack = new Stack()
myStack.push(1);
myStack.push(5);
myStack.push(76);
myStack.push(69);
myStack.push(32);
myStack.push(54);
console.log(myStack.size());
console.log(myStack.peek());
console.log(myStack.pop());
myStack.sort();
console.log(myStack.peek());
console.log(myStack.pop());
console.log(myStack.peek());
myStack.push(55);
console.log(myStack.peek());
myStack.print();
myStack.sort();
myStack.print();

File: code/data_structures/src/stack/sort_stack/sort_stack.cpp
/* Part of Cosmos by OpenGenus Foundation */

/* Sort a stack */

#include <iostream>
#include <stack>
using namespace std;
void InputElements(stack<int> &s)               //////////// Function to insert the elements in stack
{   //no. of elements to be inserted
    int size;
    cin >> size;

    while (size--)
    {
        int num;
        cin >> num;
        s.push(num);
    }
}
void PrintStack(stack<int> s)           //////////// Function to print the contents of stack
{
    while (!s.empty())
    {
        cout << s.top() << " ";
        s.pop();
    }
    cout << endl;
}
void SortedInsert(stack<int> &s, int num)  ////////// Function to insert a element at its right position in sorted way
{
    if (s.empty() || s.top() > num)
    {
        s.push(num);
        return;
    }
    int top_element = s.top();
    s.pop();
    SortedInsert(s, num);
    s.push(top_element);
}
void SortStack(stack<int> &s)            //////////// Function to sort the stack
{
    if (s.empty())
        return;
    int top_element = s.top();
    s.pop();
    SortStack(s);
    SortedInsert(s, top_element);
}
int main()
{
    stack<int> s;
    //Inserting elements in to the stack
    InputElements(s);
    //Print stack before sorting
    PrintStack(s);
    //Sort the stack
    SortStack(s);
    //Print stack after sorting
    PrintStack(s);
    return 0;
}
